get_openstack_info_open_port/get_openstack_info_open_port/get_openstack_info_open_port.py

```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/env python3
+# -*- encoding: utf-8 -*-
+#
+# This script is licensed under GNU GPL version 2.0 or above
+# (c) 2021 Antonio J. Delgado
+# __description__
+
+import sys
+import os
+import logging
+import click
+import click_config_file
+from logging.handlers import SysLogHandler
+from keystoneauth1 import session
+from keystoneauth1.identity import v3
+from keystoneclient.v3 import client as keystoneclient_v3
+from novaclient import client as nova
+from neutronclient.neutron import client as neutron
+import ipaddress
+import json
+import socket
+
+class get_openstack_info_open_port:
+
+    def __init__(self, debug_level, log_file, ip, port):
+        ''' Initial function called when object is created '''
+        self.config = dict()
+        self.config['debug_level'] = debug_level
+        if log_file is None:
+            log_file = os.path.join(os.environ.get('HOME', os.environ.get('USERPROFILE', os.getcwd())), 'log', 'get_openstack_info_open_port.log')
+        self.config['log_file'] = log_file
+        self.ip = ip
+        self.port = port
+        self._init_log()
+        self._test_open_port()
+        self._check_openstack()
+        self._os_auth()
+        self._log.debug(f"Looking for instance with IP '{ip}' in OpenStack's region '{os.environ['OS_REGION_NAME']}' as user '{os.environ['OS_USERNAME']}' with port '{port}' open to incoming connections...")
+        self._get_openstack_instances()
+        self._locate_affected_instance()
+        self._check_affected_security_groups()
+        
+    def _test_open_port(self):
+        self._log.debug(f"Testing that '{self.ip} TCP/{self.port}' is actually open...")
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        sock.settimeout(10)
+        result = sock.connect_ex((self.ip, self.port))
+        sock.close()
+        if result == 0:
+            self._log.debug(f"Port '{self.port}' is reachable.")
+            return True
+        else:
+            self._log.debug(f"Port '{self.port}' is NOT reachable.")
+            sys.exit(3)
+       
+    def _get_credentials(self):
+        """
+        Load login information from environment
+        :returns: credentials
+        :rtype: dict
+        """
+        cred = dict()
+        cred['auth_url'] = os.environ.get('OS_AUTH_URL').replace("v2.0", "v3")
```

Comment:
If just OS_AUTH_URL is missing it gives an error as it cant run the replace on a Nonetype object

Reply:
I've got this code from somewhere, I refactored it.